feature(raw): Improve prisma.raw by re-exporting all of sql-template-tag

zachasme opened the pull request:
When sending raw queries using `prisma.raw` it would be nice to have access to all of `sql-template-tag`.

We had a specific query where we wanted to write the SQL by hand. But the current API is quite limited. Nesting and joining in particular from `sql-template-tag` are very useful when composing complex queries.

This PR simply re-exports the same API as a direct import, adds a case for `prisma.sql` input to `prisma.raw`, and adds a simple test as well.

zachasme committed:
Improve Prisma.raw

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=2311) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=2311) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=2311) it.</sub>

Sytten commented:
Did you manage to get some SQL syntax highlight at the same time? This is something that I would like to see improved.

Sytten commented on src/packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -123,6 +123,14 @@ export interface GetPrismaClientOptions {
+) => sqlTemplateTag.sqltag(...args)
+sqlTemplateTagExport.empty = sqlTemplateTag.empty
+sqlTemplateTagExport.join = sqlTemplateTag.join
+sqlTemplateTagExport.raw = sqlTemplateTag.raw
```
I would prefer if we could import directly those tags from prisma instead of using the instance to reach them.

Sytten commented on src/packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -307,6 +315,12 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
+    /**
+     * Re-export of 'sql-template-tag'
+     */
+    sql = sqlTemplateTagExport
```
You need to add this in `TSClient.ts` in the folder `generation` so the typing follows.

zachasme replied:
Thanks, I think I've added the correct exports, please let me know. :-)

zachasme committed:
Export sql-template-tag directly from prisma, add typings to TSClient.ts

zachasme commented:
Thanks for the feedback, I've rebased on master, moved the exports to the module directly (instead of `prisma instance`) and added exports to `TSClient.ts`.

Sytten commented on src/packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -335,9 +336,16 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
-      } else {
+      } else if ("string" === typeof stringOrTemplateStringsArray) {
         // Called with prisma.raw(string)
         query = stringOrTemplateStringsArray
```
Here I would also add the parameters since someone could write `.raw(string with $1, values)`. This is unrelated to this PR though

zachasme replied:
Will do

Sytten commented on src/packages/client/src/runtime/getPrismaClient.ts:
```diff
@@ -335,9 +336,16 @@ export function getPrismaClient(config: GetPrismaClientOptions): any {
+        query = stringOrTemplateStringsArray[sqlOutput]
+        parameters = {
+          values: JSON.stringify(stringOrTemplateStringsArray.values),
+          __prismaRawParamaters__: true,
```
Typo here

zachasme replied:
Hm, which part? :)

Sytten replied:
Paramaters -> Parameters

zachasme replied:
Ah, this is the actual name used in the codebase.

https://github.com/prisma/prisma/blob/788f9e5b1ca024a27d24a4eb8380eb87f7b474b6/src/packages/client/src/runtime/query.ts#L684

You want this handled in another PR or this one?

Sytten replied:
I believe another PR would be better. Getting this one merged and documented would be a big win, I really need it... 
@timsuchanek @janpio

janpio replied:
Separate PR sounds better, indeed.

zachasme committed:
Add support for .raw(string with , values)

zachasme commented:
What else can I do to move this along?

timsuchanek commented:
@zachasme thanks, so far there's nothing needed from your side.
We first need to agree internally, if this is the API, that we want.
As soon as we've done so, we'll let you know here in the PR.

timsuchanek commented:
Thanks for the PR! I did a few changes about where to import stuff from, it's now building in our CI and will be available on alpha in 10 min.

Sytten commented:
Wow thanks a lot guys! Good job on landing this
